### ISSUE TITLE
Exploit in site allows for orders not fully paid to be completed

### ISSUE THREAD

**Issue by utoc** [cubecart/v6#1670]
First off, I will not post information publicly to protect those that have existing shops.  Please contact me to let me know what information you need.

Cubecart v6.1.8

Extensions:
```

Status	Name & Description	Version	Type	Developer	 
Disable	CubeCart Security Suite
The CubeCart Security Suite provides a number of tools to help secure your store.	1.0.7	Plugins	CubeCart Limited	 
Disable	Custom Foundation 5 Skin Configuration
Custom Foundation 5 Skin Configuration CC6 Plugin By SemperFiWebServices.	1.4.0	Plugins	SemperFiWebServices	 
Disable	Facebook
Facebook - comments and like buttons	1.0.0	Social	CubeCart	 
Disable	PayPal
PayPal Standard Payment Gateway	1.0.5	Gateway	CubeCart Limited
```

We had a user this morning make an order consisting of all digital downloads. Their total came out to several hundred dollars. They then somehow submitted a PayPal payment of $0.01 and their order was approved.

```

Transaction ID	Status	Amount	Gateway	Date/Time	Notes
Xxxxx	Completed	$0.01	PayPal	Today, 06:21	Payment successful. 
Address: confirmed
Payer Status: unverified This Transaction ID has been processed before.
Xxxxx	Completed	$0.01	PayPal	Today, 06:20	Payment successful. 
Address: confirmed
Payer Status: unverified
```

```

Status	Date/Time	Initiator
Failed Fraud Review	Today, 08:50	Staff
Declined	Today, 08:22	Staff
Order Complete	Today, 06:21	Gateway
Processing	Today, 06:20	Gateway
Order Complete	Today, 06:20	Gateway
Processing	Today, 06:20	Gateway
Pending	Today, 06:18	Gateway
Pending	Today, 06:18	Gateway
Pending	Today, 06:18	Gateway
Pending	Today, 05:59	Gateway
Pending	Today, 05:58	Gateway
Pending	Today, 05:58	Gateway
Pending	Today, 05:58	Gateway
Pending	Today, 05:57	Gateway
```

**Comment by abrookbanks** [comment]
No I think I got it covered thanks.

**Comment by abrookbanks** [comment]
The only was to completely protect from this is with encryption which isn't supported by PayPal. All we can do is to add one extra check. My advice is to use Express Checkout instead.

**Comment by havenswift-hosting** [comment]
Al, are you confirming that this is possible and cannot be protected against when using standard PayPal gateway module ?  If so is there any further information you can provide (publicly or privately) regarding this - this module is probably used by tens of thousands of stores and if one person has found an exploit, then others will.

**Comment by abrookbanks** [comment]
Pretty much any payment gateway that doesn't have encrypted fields can be tricked. It's easy by just editing the form fields on the transfer to gateway page.

Merchants must check payment for orders is correct  before shipping. Of course this can't be done for digital orders unless reviewed manually.

**Comment by abrookbanks** [comment]
Please use Express Checkout instead!! ☺️

**Comment by abrookbanks** [comment]
Authorize SIM is another that can be tricked.

**Comment by havenswift-hosting** [comment]
>>Pretty much any payment gateway that doesn't have encrypted fields can be tricked.  It's easy by just editing the form fields on the transfer to gateway page.

Of course !  Is it not possible to add a check just before data is sent to PayPal to check that the amounts balance against the order amount ?

>>Merchants must check payment for orders is correct before shipping.

That has always been the advice, not just for CubeCart but for any cart based system but almost nobody does and as you say, doesnt help with digital only orders anyway.

>>Authorize SIM is another that can be tricked.

I think maybe a forum post warning users against this possibility with a list of affected gateways and then a warning on each of the extension plugin pages ?

**Comment by abrookbanks** [comment]
I'll have a think about it. In the mean time I think extra protection can be added to PayPal IPN which will stop the order updating. The initial payment amount could still be changed though.

**Comment by utoc** [comment]
I just switched over to the PayPal Express Checkout module then; thats for checking into this I'll leave it to you to close out since you have some code changes.

**Comment by abrookbanks** [comment]
Before long we will have a very cool Braintree integration too.

**Comment by abrookbanks** [comment]
I released 1.0.6 which checks the paid currency and amount is correct prior to status update. 

https://www.cubecart.com/extensions/payment-gateways/paypal-standard
